File: 63.uniquePathsWithObstacles.cpp
class Solution {
public:
    int uniquePathsWithObstacles(vector<vector<int>>& obstacleGrid) {
        int m = obstacleGrid.size();
        if (m == 0) return 0;
        int n = obstacleGrid[0].size();
        if (n == 0) return 0;

        vector<vector<long>> pathCount(m, vector<long>(n, 0));
        
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if(obstacleGrid[i][j]) {
                    pathCount[i][j] = 0;
                    continue;
                }
                if (i == 0 && j == 0) pathCount[0][0] = 1;
                else {
                    if (i>0) pathCount[i][j] += pathCount[i-1][j];
                    if (j>0) pathCount[i][j] += pathCount[i][j-1];
                }
            }
        }

        return pathCount[m-1][n-1];     
    }
};

File: 63.uniquePathsWithObstacles.go
func uniquePathsWithObstacles(obstacleGrid [][]int) int {
    if obstacleGrid[0][0] == 1 {return 0}
    for i := 0; i < len(obstacleGrid); i++ {
        for j := 0; j < len(obstacleGrid[0]); j++ {
            if i == 0 && j == 0 {
                obstacleGrid[i][j] = 1
                continue
            }
            if (obstacleGrid[i][j] == 1) {
                obstacleGrid[i][j] = 0
            } else {
                obstacleGrid[i][j] = 0;
                if i >= 1 {obstacleGrid[i][j] += obstacleGrid[i-1][j]}
                if j >= 1 {obstacleGrid[i][j] += obstacleGrid[i][j-1]}
            }
        }
    }

    return obstacleGrid[len(obstacleGrid)-1][len(obstacleGrid[0])-1]
}